samza-core/src/main/java/org/apache/samza/operators/impl/OperatorSpecGraph.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.operators.impl;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.samza.SamzaException;
+import org.apache.samza.operators.ContextManager;
+import org.apache.samza.operators.StreamGraphImpl;
+import org.apache.samza.operators.spec.InputOperatorSpec;
+import org.apache.samza.operators.spec.OperatorSpec;
+import org.apache.samza.operators.spec.OutputStreamImpl;
+import org.apache.samza.system.StreamSpec;
+
+import static com.google.common.base.Preconditions.*;
+
+
+/**
+ * Defines the serialized format of {@link StreamGraphImpl}. This class encapsulates all getter methods to get the {@link OperatorSpec}
+ * initialized in the {@link StreamGraphImpl} and constructsthe corresponding serialized instances of {@link OperatorSpec}.
+ * The {@link StreamGraphImpl} and {@link OperatorSpec} instances included in this class are considered as immutable and read-only.
+ * The instance of {@link OperatorSpecGraph} should only be used in runtime to construct {@link org.apache.samza.task.StreamOperatorTask}.
+ */
+public class OperatorSpecGraph {
+  private final Map<String, OperatorSpec> operatorSpecMap = new HashMap<>();
+  private final Map<String, byte[]> serializedOpSpecs = new ConcurrentHashMap<>();
+  private final StreamGraphImpl originalGraph;
+
+  public OperatorSpecGraph(StreamGraphImpl streamGraph) {
```

Comment:
Discussed offline: 
If we serialize this graph (both the operator specs vertices and the edges) in a planner, we can then avoid having to re-init the StreamApplication at runtime in the container to get the StreamGraphImpl (which is actually OperatorSpecGraph). This will also makes it easier to understand the OperatorImplGraph - it should take the serialized representation of the graph of operator specs, and create a corresponding graph of OperatorImpls.